diff --git toolchain/internal/common.bzl toolchain/internal/common.bzl
index bb2fcfa..eacd236 100644
--- toolchain/internal/common.bzl
+++ toolchain/internal/common.bzl
@@ -60,7 +60,8 @@ def _linux_dist(rctx):
     return distname, version
 
 def os_version_arch(rctx):
     _os = os(rctx)
     _arch = arch(rctx)
+    return _os, "", _arch
 
     if _os == "linux":
@@ -72,6 +72,8 @@ def os_version_arch(rctx):
     return _os, "", _arch
 
 def os(rctx):
+    if rctx.attr.exec_os:
+        return rctx.attr.exec_os
     # Less granular host OS name, e.g. linux.
 
     name = rctx.os.name
@@ -88,6 +90,8 @@ def os_bzl(os):
     return {"darwin": "osx", "linux": "linux"}[os]
 
 def arch(rctx):
+    if rctx.attr.exec_cpu:
+        return rctx.attr.exec_cpu
     arch = rctx.os.arch
     if arch == "arm64":
         return "aarch64"
diff --git toolchain/internal/repo.bzl toolchain/internal/repo.bzl
index 67b3ed82d..02bdfc812 100644
--- toolchain/internal/repo.bzl
+++ toolchain/internal/repo.bzl
@@ -35,6 +35,14 @@ common_attrs = {
                "in the list of known llvm_distributions using the provided version. " +
                "If unset, a default value is set from the `llvm_version` attribute."),
     ),
+    "exec_os": attr.string(
+        mandatory = True,
+        doc = "exec os for the toolchain, default to host os",
+    ),
+    "exec_cpu": attr.string(
+        mandatory = True,
+        doc = "exec cpu for the toolchain, default to host arch",
+    ),
 }
 
 llvm_repo_attrs = dict(common_attrs)
