diff --git deps.bzl deps.bzl
index 5570688..844e959 100644
--- deps.bzl
+++ deps.bzl
@@ -767,5 +767,5 @@ def bazeldnf_dependencies():
     )
 
 def _maybe(name, **kwargs):
-    if name not in native.existing_rules():
+    if name not in native.existing_rules() and ("org_golang" not in name or "org_golang_x_crypto" == name) and "prometheus" not in name:
         _go_repository(name = name, **kwargs)
